### PR TITLE
Adds DNS resolver configuration option for ApacheHttpClient

### PR DESCRIPTION
**Pull request by cenedhryn**
## Motivation and Context
Implements a feature requested in [#2627](https://github.com/aws/aws-sdk-java-v2/issues/2627)

## Description
Allows users to set a custom DNS resolver to be used for the Apache Http Client.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: .changes/next-release/feature-AWSSDKforJavav2-418a15a.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "category": "AWS SDK for Java v2",
```
Comment (dagnir) [inline .changes/next-release/feature-AWSSDKforJavav2-418a15a.json:2]: Since this is specific to Apache, we should change this to the Apache client

File: http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientWireMockTest.java [hunk 2]
```diff
@@ -148,4 +158,52 @@ public void clear() {
...
+    public void overrideDnsResolver_withDnsMatchingResolver_successful() throws Exception {
+        overrideDnsResolver("magic.local.host");
+    }
+
+    @Test(expected = UnknownHostException.class)
+    public void overrideDnsResolver_withLocalhostIp_successful() throws Exception {
+        overrideDnsResolver("sad.local.host");
+    }
+
+    private void overrideDnsResolver(String hostName) throws IOException {
```
Comment (dagnir) [inline http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientWireMockTest.java:172]: `hostName` looks unused

File: http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientTest.java [hunk 3]
```diff
@@ -123,4 +126,23 @@ public void credentialProviderCanBeUsedWithProxy() {
                         .credentialsProvider(Mockito.mock(CredentialsProvider.class))
                         .build();
     }
+
+    @Test
+    public void dnsResolverCanBeUsed() {
```
Comment (dagnir) [inline http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientTest.java:131]: can we add a test that if the caller explicitly sets to null we use the default resolver?

File: .changes/next-release/feature-AWSSDKforJavav2-418a15a.json (hunk 1 above)
Comment (cenedhryn) [reply .changes/next-release/feature-AWSSDKforJavav2-418a15a.json:2]: ok.

File: http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientTest.java (hunk 3 above)
Comment (cenedhryn) [reply http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientTest.java:131]: yes

File: http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientWireMockTest.java (hunk 2 above)
Comment (cenedhryn) [reply http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientWireMockTest.java:172]: will take a look

Comment (cenedhryn) [reply http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientWireMockTest.java:182]: It's used for the URI a bit further down
